### PR TITLE
Submission fixes

### PR DESCRIPTION
**Pull request by jannisborn**
- Miscellaneous fixes in notebooks for submission
- improved error handling and documentation in RT
- fixed a bug related to GPU usage in paccmann_gp inference
- improved error handling in `algorithms/core.py`

### CODE REVIEW

File: src/gt4sd/algorithms/core.py [hunk 1]
```diff
@@ -57,7 +57,7 @@
     sync_algorithm_with_s3,
     upload_to_s3,
 )
-from ..exceptions import InvalidItem, S3SyncError, SamplingError
+from ..exceptions import Gt4sdTimeoutError, InvalidItem, S3SyncError, SamplingError
```
Comment (drugilsberg) [inline src/gt4sd/algorithms/core.py:60]:
```suggestion
from ..exceptions import GT4SDTimeoutError, InvalidItem, S3SyncError, SamplingError
```

File: src/gt4sd/algorithms/core.py [hunk 2]
```diff
@@ -166,7 +166,7 @@ def _setup_untargeted_generator(
             return partial(self.generator, self.target)  # type: ignore
 
     def timeout(self, signum, frame):
-        raise TimeoutError(
+        raise Gt4sdTimeoutError(
```
Comment (drugilsberg) [inline src/gt4sd/algorithms/core.py:169]:
```suggestion
        raise GT4SDTimeoutError(
```

File: src/gt4sd/algorithms/core.py [hunk 3]
```diff
@@ -265,8 +265,8 @@ def raise_if_none_sampled(items: set, detail: str):
                     raise_if_none_sampled(items=item_set, detail=detail)
                     return
                 item_set_length = len(item_set)
-        except TimeoutError:
-            detail = f"Samples took longer than {self.max_runtime} seconds to generate."
+        except Gt4sdTimeoutError:
```
Comment (drugilsberg) [inline src/gt4sd/algorithms/core.py:268]:
```suggestion
        except GT4SDTimeoutError:
```

File: src/gt4sd/exceptions.py [hunk 4]
```diff
@@ -102,3 +102,18 @@ def __init__(self, title: str, detail: str) -> None:
         self.title = title
         self.detail = detail
         super().__init__(detail)
+
+
+class Gt4sdTimeoutError(TimeoutError):
```
Comment (drugilsberg) [inline src/gt4sd/exceptions.py:107]:
```suggestion
class GT4SDTimeoutError(TimeoutError):
```

**Review by drugilsberg — CHANGES_REQUESTED** [review]
Great job, only minor changes

**Review by drugilsberg — COMMENTED** [review]
Looks good, I would consider cleaning the notebooks output before merging.

**Comment by jannisborn** [comment]
We need this additional exception class because `SamplingError` is a child of the native `TimeoutError` and thus the `except TimeoutError:` block is entered not only if a `TimeoutError` has been raised but also if a `SamplingError` was raised. This led to wrong error messages

**Comment by drugilsberg** [comment]
> We need this additional exception class because `SamplingError` is a child of the native `TimeoutError` and thus the `except TimeoutError:` block is entered not only if a `TimeoutError` has been raised but also if a `SamplingError` was raised. This led to wrong error messages

Makes sense, this was a good catch.

**Review by drugilsberg — APPROVED** [review]

**Comment by jannisborn** [comment]
New improvements:
[refactor: remove .core from inits](https://github.com/GT4SD/gt4sd-core/pull/167/commits/8994a421bfc430a9294a518c9aee5c80edaf968e)
[fix: torchdrug openMP mismatch](https://github.com/GT4SD/gt4sd-core/pull/167/commits/9d61854f9bdbd6f3b6e3c54a95606cff4d4003b0)
[refactor: avoid signal to run core outside main thread](https://github.com/GT4SD/gt4sd-core/pull/167/commits/3ef6f4d7b521738c8575981b959a2f7b3911d038)

File: src/gt4sd/algorithms/core.py [hunk 5]
```diff
@@ -141,6 +141,28 @@ def get_generator(
             If the target is None, the generator is assumed to be untargeted.
         """
 
+    def timeout(self, item_set: Set, detail: str, error: TimeoutError):
```
Comment (drugilsberg) [inline src/gt4sd/algorithms/core.py:144]:
```suggestion
    def timeout(self, item_set: Set, detail: str, error: TimeoutError) -> None:
```

File: src/gt4sd/algorithms/core.py [hunk 6]
```diff
@@ -141,6 +141,28 @@ def get_generator(
...
+
+        Args:
+            item_set: Set of items generated thus far.
+            detail: context or condition for the generation. Defaults to None.
+            error: An error instance, child class of `TimeoutError` either
+                `GT4SDTimeoutError` or `SamplingError`.
+
+        Raises:
+            TimeoutError: If no items were sampled so far.
+
```
Comment (drugilsberg) [inline src/gt4sd/algorithms/core.py:156]:
```suggestion
```

**Review by drugilsberg — APPROVED** [review]
Looks great only minor suggestions for typing and docstring in the timeout function.

File: src/gt4sd/algorithms/core.py [hunk 7]
```diff
@@ -141,6 +141,28 @@ def get_generator(
             If the target is None, the generator is assumed to be untargeted.
         """
 
+    def timeout(self, item_set: Set, detail: str, error: TimeoutError):
+        """Throws a timeout exception if applicable, otherwise returns
+            items gracefully.
+
+        Args:
+            item_set: Set of items generated thus far.
+            detail: context or condition for the generation. Defaults to None.
```
Comment (jannisborn) [inline src/gt4sd/algorithms/core.py:150]:
```suggestion
            detail: context or condition for the generation.
```

**Pull request merged**